Add more error handling in writeBytes.

hiddenalpha opened the pull request:
- Throw NullPointerException (in place of crashing) if caller passes null buffer.
- Pass-through OutOfMemoryError if GetByteArrayElements() fails to alloc memory.

hiddenalpha committed:
Add more error handling in writeBytes.

- Throw NullPointerException (in place of crashing) if caller passes
  null buffer.
- Pass-through OutOfMemoryError if GetByteArrayElements() fails to alloc
  memory.

Cherry-picked from 5abdd451099e5749b92afc8fc189e8afca2985a2.

tresf commented:
> Pass-through OutOfMemoryError if GetByteArrayElements() fails to alloc memory.

Can you explain where the `OutOfMemoryError` is thrown?

> Throw NullPointerException (in place of crashing) if caller passes null buffer.

Hmm... Should we wrapped exceptions which are historically non-checked?

hiddenalpha committed:
I was really surprised to find that GetByteArrayElements seems to just
return NULL, without any code that would prepare a java exception. In
JNI doc I read that an exception gets setup if a JNI function fails.
(Read "https://www.developer.com/open-source/exception-handling-in-jni/"
to see what I mean) But for whatever reason this seems not to be the
case for GetByteArrayElements().

- writeBytes() calls GetByteArrayElements().
- GetByteArrayElements() calls NEW_C_HEAP_ARRAY_RETURN_NULL().
- NEW_C_HEAP_ARRAY_RETURN_NULL() calls NEW_C_HEAP_ARRAY3().
- NEW_C_HEAP_ARRAY3() calls AllocateHeap().
- AllocateHeap() uses malloc() to get memory.

Direct Links:
GetByteArrayElements: https://github.com/openjdk/jdk/blob/jdk8-b120/hotspot/src/share/vm/prims/jni.cpp#L3597
call to NEW_C_HEAP_ARRAY_RETURN_NULL: https://github.com/openjdk/jdk/blob/jdk8-b120/hotspot/src/share/vm/prims/jni.cpp#L3611
NEW_C_HEAP_ARRAY_RETURN_NULL: https://github.com/openjdk/jdk/blob/jdk8-b120/hotspot/src/share/vm/memory/allocation.hpp#L680
call to NEW_C_HEAP_ARRAY3: https://github.com/openjdk/jdk/blob/jdk8-b120/hotspot/src/share/vm/memory/allocation.hpp#L681
NEW_C_HEAP_ARRAY3: https://github.com/openjdk/jdk/blob/jdk8-b120/hotspot/src/share/vm/memory/allocation.hpp#L668
call to AllocateHeap: https://github.com/openjdk/jdk/blob/jdk8-b120/hotspot/src/share/vm/memory/allocation.hpp#L669
AllocateHeap: https://github.com/openjdk/jdk/blob/jdk8-b120/hotspot/src/share/vm/memory/allocation.inline.hpp#L52

Picked from 4a805a7902cf0c572f7237661f2684cf9f557f0b.

hiddenalpha commented:
Will have a look as soon I find some time. Thx for feedback so far.

tresf commented:
> > Hmm... Should we wrap exceptions which are historically non-checked?

> Will have a look as soon I find some time. Thx for feedback so far.

Whoops, we don't.  We only catch `IOException`, so we're still going to throw unchecked exceptions, which I think is what we want, per:

https://github.com/java-native/jssc/blob/ea2ade2b3cbbc417ef9dfd97922a77fde41b911a/src/main/java/jssc/SerialPort.java#L414-L416

I think unchecked exceptions are fine for now (they're certainly better than a JVM crash).  If users complain about this we, can modify the the catch statement.

tresf commented:
@hiddenalpha do you believe we should add more unit tests for these?  Furthemore, have you had time to evaluate the safety of the windows code too?

hiddenalpha commented:
> do you believe we should add more unit tests for these?

I think writing unit tests may be challenging for some of those cases. Especially edge-cases in the native code. I think best is to add a test when we see a good opportunity and if it can be written in a simple (*as in understandable/maintainable*) way.

> have you had time to evaluate the safety of the windows code too?

Did not look at the windows code at all yet. Will keep it in mind when I've a free time slot.

hiddenalpha committed:
Add same null checks for windoof. Add unit test for the null buffer case

Picked from 86aad7e512ad3b9d1f551286bc4dc8d2413720ab.

tresf commented on src/test/java/jssc/SerialNativeInterfaceTest.java:
```diff
@@ -47,4 +55,16 @@ public void reportsWriteErrorsAsIOException() throws Exception {
         testTarget.writeBytes(fd, buf);
     }
 
+    @Test
+    public void throwsNpeIfPassedBufferIsNull() throws Exception {
+        long fd = 1;
+        byte[] buf = null;
+        try{
+            testTarget.writeBytes(fd, buf);
+            fail("Where is the exception?");
+        }catch( NullPointerException ex ){
+            assertTrue(ex.getMessage().contains("buffer"));
+        }
+    }
```
Should we copy the `expected =` signature in case we change the messaging later?

```suggestion
    @Test(expected = java.lang.NullPointerException.class)
    public void throwsNpeIfPassedBufferIsNull() throws Exception {
        long fd = 1;
        byte[] buf = null;
        testTarget.writeBytes(fd, buf);
    }
```

tresf commented:
This can be tested in its current form using `mvn -DskipTests install && mvn` until rebased against master (specifically #51, #151).

It looks good to me, but I've left a question about the unit test, since matching an exception string can have unobvious test failures in the even this ever gets changed.

hiddenalpha committed:
Merge remote-tracking branch 'upstream/master' into more-error-handling-in-writeBytes-20231109

Conflicts:
  src/test/java/jssc/SerialNativeInterfaceTest.java

tresf commented:
When you have time, this PR needs to be rebased and there's still one comment remaining.  I've love to see it merged. :)

hiddenalpha committed:
Apply https://github.com/java-native/jssc/pull/152#discussion_r1394476235

Taken from 618158eb48bbf545cef4cb449cdad3c87e9a6775 .
